taichi/program/state_flow_graph.cpp

```diff
@@ -197,6 +211,15 @@ bool StateToNodesMap::StateIterator::has_edge(Node *n) const {
   return std::binary_search(cur_, end_, e);
 }
 
+// static
+int StateToNodesMap::get_mask_slot(const Edge &e) {
+  constexpr int kNumBits = sizeof(MaskType) * 8;
+  // Divided by 8 to avoid trailing zeros.
+  const size_t h =
+      e.first.unique_id ^ (reinterpret_cast<uintptr_t>(e.second) / 8);
```

Comment:
I'm not sure if `e.second->node_id` is a better key than `e.second`: the former is more compact yet leads to one more load. We can probably decide using a benchmark via https://github.com/taichi-dev/taichi/pull/2090 :-)

Reply:
Ah ok, left a TODO..